src/graph.jl

```diff
@@ -0,0 +1,349 @@
+
+    
+    vertex_types = CSV.read(vertex_type_defs)
+    
+    synth_vertices_df = []
+    
+    for vertex in eachrow(vertex_types)
+        
+        # Generate a synthetic name; hash the concatenation of name||type to get a unique vertex id
+        v_name = string("synth_", vertex.v_type, "_", randstring(MersenneTwister(), 'a':'z', 6)) 
+        v_hash = hash(string(v_name, vertex.v_type))       
+        synth_v_attrs = NamedTuple{(:v_hash, :v_name, :v_type)}(("$v_hash", v_name, vertex.v_type))
+        
+        push!(synth_vertices_df, synth_v_attrs)
+        
+    end
+    
+    open(output_path, "w") do io
+        print(io, repr(load_graph_data(synth_vertices_df)))
+    end
+
+    #CSV.write(output_path, synth_vertices_df)
+    @info "Synthetic vertex dataframe generated and saved as a Julia file."
+    
+    return DataFrame(synth_vertices_df)
+
+end
+
+
+function generate_synthetic_edges(edge_type_defs::String, synth_vertex_df::DataFrame, output_path::String)
+    
+    """
+    This function ingests a synthetic vertex data frame and generates synthetic edges that are consistent with edge src/dst vertex type constraints. 
+    This synthetic vertex dataframe can be used for testing/debugging/developing the knowledge graph.
+    """
+    
+    edge_types = CSV.read(edge_type_defs)
+
+    synth_edges_df = []
+    
+    for edge_row in eachrow(edge_types)
+
+        src_row = @linq synth_vertex_df |> 
+            where(:v_type .== edge_row.src_type) |> 
+            select(:v_hash, :v_name, :v_type)
+        
+        dst_row = @linq synth_vertex_df |> 
+            where(:v_type .== edge_row.dst_type) |>
+            select(:v_hash, :v_name, :v_type)
+        
+        if size(src_row)[1] >=1 && size(dst_row)[1] >= 1
+            src_vhash = src_row.v_hash[1]
+            src_name = src_row.v_name[1]
+            src_vtype = src_row.v_type[1]
+
+            dst_vhash = dst_row.v_hash[1]
+            dst_name = dst_row.v_name[1]
+            dst_vtype = dst_row.v_type[1]
+            
+            if edge_row.value_field != "nothing"
+                edge_val = ("placeholder")
+            else
+                edge_val = nothing
+            end
+
+            synth_edge_attrs = NamedTuple{(:src_vhash, :src_name, :src_vtype, :dst_vhash, :dst_name, :dst_vtype, :edge_relation, :edge_description, :value)}(("$src_vhash", src_name, src_vtype, "$dst_vhash", dst_name, dst_vtype, edge_row.edge_relation, edge_row.description, edge_val))
+            push!(synth_edges_df, synth_edge_attrs)
+            
+        else
+            continue
+        end
+    end
+    
+    open(output_path, "w") do io
+        print(io, repr(load_graph_data(synth_edges_df)))
+    end
+
+    #CSV.write(output_path, synth_edges_df)
+    @info "Synthetic edge dataframe generated and saved as a Julia file."
+    
+    return DataFrame(synth_edges_df)
+
+end
+
+function insert_vertices_from_jl(vertices_file::String, input_graph::Union{MetaDiGraph, Nothing})
```

Comment:
I think the standard way to do optional arguments is either to have a default arguments like:
    `function insert_vertices_from_jl(vertices_file::String, input_graph=MetaDiGraph())`

Or to have two functions, one that does the modifying and one that creates the "default input"
    `function insert_vertices_from_jl!(vertices_file::String, input_graph::MetaDiGraph)` and 

```julia
function vertices_from_jl(vertices_file::String) 
    g = MetaDiGraph()
    insert_vertices_from_jl!(vertices_file, g)
    return g
end
```

The `Union{T,Nothing}` works, but is less idiomatic than the others. Whenever you have a union, you are creating an opportunity to bring type system logic into the runtime. The "small union optimization" makes that feasible, but there are ways to avoid it with default args and the two function approach shown above.

Reply:
I think creating two functions is probably the way to go; I figured the union approach was a bit hacky but would work for v0. I'll overload the `insert_vertices` function so each case can be handled separately.